File: replace_token/PytorchInvocation/test_ops_custom.cpp
#include <torch/torch.h>
#include <torch/script.h>
#include <ATen/npu/NPUBridge.h>
#include <iostream>

// 自定义的replace_token操作声明
torch::Tensor npu_replace_token(const torch::Tensor& x, const torch::Tensor& y);

// 测试类
class TestCustomAdd {
public:
    void test_add_custom() {
        try {
            // 设置NPU设备
            auto device = torch::Device("npu:0");
            
            // 创建输入张量
            auto options = torch::TensorOptions()
                .dtype(torch::kInt32)
                .device(torch::kCPU);
                
            // 生成随机张量，大小为(5,1)
            auto x = torch::randint(-4, 5, {5, 1}, options);
            auto y = torch::randint(1, 5, {5, 1}, options);
            
            // 打印输入张量
            std::cout << "Input x:\n" << x << std::endl;
            std::cout << "Input y:\n" << y << std::endl;
            
            // 将张量移动到NPU设备
            auto x_npu = x.to(device);
            auto y_npu = y.to(device);
            
            // 设置不需要梯度
            x_npu.requires_grad_(false);
            y_npu.requires_grad_(false);
            
            // 执行replace_token操作
            auto output_npu = torch::ops::custom_ops::replace_token(x_npu, y_npu);
            
            // 同步NPU操作
            c10::npu::NPUStream::getCurrentStream().synchronize();
            
            // 将结果移回CPU并打印
            auto output = output_npu.to(torch::kCPU);
            std::cout << "Output:\n" << output << std::endl;
            
        } catch (const std::exception& e) {
            std::cerr << "Error: " << e.what() << std::endl;
        }
    }
};

int main() {
    // 初始化NPU设备
    torch::npu::init_npu();
    
    // 禁用内部格式
    torch::npu::config::allow_internal_format = false;
    
    TestCustomAdd test;
    test.test_add_custom();
    
    return 0;
} 

File: replace_token/PytorchInvocation/test_ops_custom.py
#!/usr/bin/python3
# coding=utf-8
#
# Copyright (C) 2023-2024. Huawei Technologies Co., Ltd. All rights reserved.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# ===============================================================================

import torch
import torch_npu
# from torch_npu.testing.testcase import TestCase, run_tests

torch.npu.config.allow_internal_format = False


class TestCustomAdd():

    def test_add_custom(self):
        size = (5, 1)  # 直接用元组表示尺寸
        x = torch.randint(-4, 5, size, device='cpu', dtype=torch.int32)  # 区间[-4, 4]
        y = torch.randint(1, 5, size, device='cpu', dtype=torch.int32)
        print(x, '\n', y)

        torch.npu.synchronize()
        output = torch_npu.npu_replace_token(x.npu(), y.npu()).cpu()
        torch.npu.synchronize()

        print(output)
        # self.assertRtolEqual(output, x + y)


if __name__ == "__main__":
    # run_tests()
    test = TestCustomAdd()
    test.test_add_custom()
